### PR TITLE
[network_graph] multi output

### LINKED ISSUE #1367: Multi outputs
- The list of output layer will be maintained by graph_core.
- Switch order of topological sort and add loss layer function

### PR DESCRIPTION
**Pull request by lhs8928**
Commit 1:
    [graph_core] Implement input_list, output_list for multi input, output
    
    - Make input_list, output_list and its getter to support multi input, output
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 2:
    [network_graph] revise addlosslayer for multi output
    
     - reorder addlosslayer to run before topologicalsort
     - Replace node_list rather than delete when the loss is entropy
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 3:
    [unittest] multi output unittest
    
     - Modify unittest to test multi output case
     - Make setLabel function in neuralnet
       to set labels at all loss layers
    
    Resolve #1367
    
    **Self evaluation:**
    1. Build test: [X]Passed [ ]Failed [ ]Skipped
    2. Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/graph/graph_core.h
Comment (kparichay) [inline nntrainer/graph/graph_core.h]: I wonder if the graph_core needs this function? can graph_core extract this information by itself? I recommend to modify the modify graph core interface for graph core to interpret itself given the nodes, which are input nodes and which are output nodes.

Comment (kparichay) [reply nntrainer/graph/graph_core.h]: you might have to introduce a function (getOutputConnections() like getInputConnections()) in GraphNode.

**Comment by kparichay** [comment]
Overall good job

File: nntrainer/graph/network_graph.cpp
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:451]: Could you explain why we need to check it is bipartite?

File: nntrainer/graph/graph_core.h
Comment (zhoonit) [inline nntrainer/graph/graph_core.h:205]: Please add doxygen for the functions added.

File: nntrainer/models/neuralnet.cpp
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:487]: Currently, this is returning vector of vector.
The reasoning behind this is that each node can have multiple outputs.
eg)
out_layer1 -> out_tensor_a, out_tensor_b
out_layer2 -> out_tensor_c, out_tensor_d
out_layer3 -> out_tensor_e,

so returning [[out_tensor_a, out_tensor_b], [out_tensor_c, out_tensor_d], [out_tensor_e]]

We made a bit of offline discussion about this signature and rollback.

Reasoning,
1. NNStreamer or other frameworks expects/makes output of 1d list. (correct me if I am wrong)
2. If we start to allow nested output, not only 2d but nd list is possible. eg) multiple subgraph of above example
3. When receiving inputs, we are having 1D list (float **) not (float ***) so it is also predictable that output is also a type of 1D list.
3. Even though returning 1D result, we are not loosing much representational power because user will know about the output layout.

-------------- Off-the-topic issue
As our sort order is not determined, it will be requested to user to write the out layer names in order, to determine the output layout.

File: nntrainer/graph/network_graph.cpp
Comment (lhs8928) [reply nntrainer/graph/network_graph.cpp:451]: Not related to this pr, but when the multi input and multi output is supported the graph can be bipartite so we need to check it I guess.

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h]: So you means instread of addInputNode and addOutputNode which is called in realizeGraph by network_graph make a function in graph_core something like findinputOutputNode and call that function after running the realizeGraph. And getOutputConnection will be needed when making that function am I right?

Comment (kparichay) [reply nntrainer/graph/graph_core.h]: yes, you are right.

File: nntrainer/graph/network_graph.cpp
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:451]: bipartite graph can be cyclic (bipartite only ensures no cycles of odd length), so you might want to expand on this comment here.

File: nntrainer/graph/graph_core.h
Comment (kparichay) [reply nntrainer/graph/graph_core.h:205]: I wonder how doxygen CI passed for this PR.

File: nntrainer/models/neuralnet.cpp
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:487]: > As our sort order is not determined, it will be requested to user to write the out layer names in order, to determine the output layout.

If these names are to be taken, it better to take them during the model compile. In this way, the shape and size of every inference remains fixed.

Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:487]: > We made a bit of offline discussion about this signature and rollback.

Keeping the current signature must be clarified - does it mean multiple output nodes are allowed when the output of each node has to be just 1 tensor or a single node with multiple outputs? IMO, the former case is better, as this can be expanded by making the tensor class more flexible (supporting more than 4 dimensions in the future).

> NNStreamer or other frameworks expects/makes output of 1d list. (correct me if I am wrong)

You are right. However IMO nntrainer's tensor filter element can define semantics for the conversion from nntrainer's inference to be returned to nnstreamer.

> If we start to allow nested output, not only 2d but nd list is possible. eg) multiple subgraph of above example

Can this be nd list (as long as a node can only handle list of tensors)?

Comment (zhoonit) [reply nntrainer/models/neuralnet.cpp:487]: >  IMO, the former case is better 

I agree :)

> Can this be nd list (as long as a node can only handle list of tensors)?

Yes,

for example.

subgraph 1. [[a], [b,c], [d,e]]
subgarph 2. [[f], [g, h]

output of sugraph 1 + subgraph 2 is 3D list.

This implies that we can't fully represent the list of list with nd graph where n is a fixed number. So the best we can do would be have semantics that is is always flattened for the model output.

File: nntrainer/graph/network_graph.cpp
Comment (lhs8928) [reply nntrainer/graph/network_graph.cpp:451]: Oh no sorry. I mean 2 separated graph not the bipartite. I will amend the comment

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h]: I've done what you said but check replaceOutputNode function that this also needs to removed. Actually we can also delete this function by realize input output node after addLossLayer. But in this case in the we didn't have a information which layer is a output node so we need to check all the layer to find output layer in addLossLayer

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h:206]: Added doxygen

File: nntrainer/models/neuralnet.cpp
Comment (lhs8928) [reply nntrainer/models/neuralnet.cpp]: reverted to return sharedtensors

File: nntrainer/graph/graph_core.cpp
Comment (zhoonit) [inline nntrainer/graph/graph_core.cpp:179]: Please note that this will have to change after merging v2. it will compile so it will be hard to find the problem.

File: nntrainer/graph/graph_core.h
Comment (zhoonit) [inline nntrainer/graph/graph_core.h:198]: This has missing doxygen

File: test/unittest/unittest_nntrainer_models.cpp
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_models.cpp:1296]: is this considered resnet_like?

**Review by zhoonit — APPROVED** [review]
Overall good job! lgtm after ci pass + doxygen added

File: nntrainer/graph/graph_core.cpp
Comment (kparichay) [reply nntrainer/graph/graph_core.cpp:179]: yes, this has been added to the list of manual merging along with padding :)

Comment (kparichay) [inline nntrainer/graph/graph_core.cpp:179]: can you also add a check here ensuring that `iter->getInputConnections().size() == 1`?

File: nntrainer/graph/graph_core.cpp
Comment (kparichay) [inline nntrainer/graph/graph_core.cpp:182]: same here.

File: nntrainer/graph/graph_core.h
Comment (kparichay) [inline nntrainer/graph/graph_core.h:219]: getInputNodeSize() -> getNumInputNodes()?

File: nntrainer/graph/graph_core.h
Comment (kparichay) [inline nntrainer/graph/graph_core.h:234]: getNumOutputsNodes()

File: test/unittest/unittest_nntrainer_models.cpp
Comment (kparichay) [reply test/unittest/unittest_nntrainer_models.cpp:1296]: maybe rename to multiple output model?

**Review by kparichay — APPROVED** [review]
LGTM other than some minor comments
Amazing Job

File: nntrainer/graph/graph_core.h
Comment (kparichay) [reply nntrainer/graph/graph_core.h]: Yes, that will be good. `addLossLayer()` can once iterate over all the nodes to find output layers - it's fine IMO.
It's an optimization that can be done in the next PR as this PR is already big enough.

File: nntrainer/graph/network_graph.cpp
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp]: yes, thanks for the clarification.

File: nntrainer/models/neuralnet.cpp
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp]: great.

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h:203]: Done

File: test/unittest/unittest_nntrainer_models.cpp
Comment (lhs8928) [reply test/unittest/unittest_nntrainer_models.cpp]: Rename to multiple_output_model

File: nntrainer/graph/graph_core.cpp
Comment (lhs8928) [reply nntrainer/graph/graph_core.cpp]: Done

File: nntrainer/graph/graph_core.cpp
Comment (lhs8928) [reply nntrainer/graph/graph_core.cpp]: Done

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h]: Done

File: nntrainer/graph/graph_core.h
Comment (lhs8928) [reply nntrainer/graph/graph_core.h]: Done

File: nntrainer/graph/network_graph.cpp
Comment (kparichay) [inline nntrainer/graph/network_graph.cpp:224]: so, if some loss layers are added as layers in the model, then this continue will lead to `loss` being added after all the output nodes which do not contain loss nodes. Is this intended?

Comment (kparichay) [inline nntrainer/graph/network_graph.cpp:224]: also please add a todo here to realize the cross_entropy loss to softmax/sigmoid cross entropy loss.
This is already implemented in layer_v2, and will be added here upon merge.

Comment (lhs8928) [reply nntrainer/graph/network_graph.cpp:224]: I think you want to say that in that case it makes an error and that's what we've discussed. In this pr I didn't handle such cases. Maybe in later pr I will take care of.

Comment (lhs8928) [reply nntrainer/graph/network_graph.cpp:224]: I think from line 235 to 249 the realization of cross_entropy loss to softmax/sigmoid cross entropy loss exist. Am I misunderstand what you say?

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:224]: cool.

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:224]: If the output_layer_node is cross_entropy layer type, this loop will continue here. However, it also needs realization (not add new loss layer but udpate current loss layer type).

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/graph/network_graph.cpp
Comment (lhs8928) [reply nntrainer/graph/network_graph.cpp:224]: Oh now I understand. Thanks

File: nntrainer/graph/graph_core.h
Comment (jijoongmoon) [inline nntrainer/graph/graph_core.h:243]: I wonder if the size of ouptut_list is the same with graph node size according to line #293

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**
